beginner_source/hybrid_frontend/introduction_to_hybrid_frontend_tutorial.py

```diff
@@ -0,0 +1,526 @@
+# write a model definition in Python, which is run later on in the Caffe2
+# C++ runtime environment. Having a deferred representation has numerous
+# benefits, including:
+#
+#   - *Model simplicity*: Model representation is simpler and more restricted
+#   - *Optimization*: We can apply various optimization techniques to create
+#     more efficient deployable models
+#   - *Improved performance*: We can leverage parallel and out-of-order
+#     execution, and target highly optimized hardware architectures
+#   - *Easy export*: Exporting a serialized model is a breeze with ONNX
+#
+# .. figure:: /_static/img/hybrid_frontend/pytorch_workflow_small.jpg
+#    :alt: workflow
+
+
+
+
+######################################################################
+# JIT Compiler
+# ------------
+#
+# To accomodate to as many workflows as possible, PyTorch not only
+# provides both an eager mode and a graph mode, but a transition mechanism
+# between them: ``torch.jit``. This enables the use of the newest
+# state-of-the-art models from research environments to be deployed and
+# exported in demanding production environments without massive code
+# re-writes.
+#
+# To learn more see `this
+# letter <https://pytorch.org/2018/05/02/road-to-1.0.html>`__ from the
+# PyTorch team.
+#
+# There are two separate modes for performing this transition.
+#
+# Tracing Mode
+# ~~~~~~~~~~~~
+#
+# Tracing mode is a non-invasive tool which traces your Python module or
+# function to create an internal representation that is compiled to graph
+# mode. The ``torch.jit.trace`` function takes your module or function and
+# an example data input, and traces the computational steps that the data
+# encounters as it progresses through the model.
+#
+# An advantage of this approach is that it is an unobtrusive process that
+# will work regardless of how your Python code is structured. However,
+# because of the way the tracer works, data-dependent control flows (if
+# statements, loops) cannot be fully captured. Rather, only the control
+# sequence traced using the example input will be recorded.
+#
+# Script Mode
+# ~~~~~~~~~~~
+#
+# To compile models with data-dependent control flow elements, such as
+# RNNs, we simply add a ``@torch.jit.script`` decorator to our function.
+# This annotation lets PyTorch know that the function may contain
+# data-dependent control flows. Now, our loops, conditionals, and
+# iterative data mutations are handled explicitly.
+#
+# The one constraint for users compiling with script mode is that it
+# currently only supports a restricted subset of Python. As of now,
+# features such as generators, defs, and Python data structures are not
+# supported. To remedy this, you can invoke traced modules from script
+# modules (and vice-versa), and you can call pure Python functions and
+# modules from script modules. However, the operations done in the pure
+# Python functions will not be compiled, and will run as-is.
```

Comment:
"will not be exportable" <- this is probably the biggest distinction here since if you want to run the model in production you will need to remove all the python calls.